src/xrpld/app/tx/detail/DelegateSet.cpp

```diff
@@ -0,0 +1,159 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2025 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <xrpld/app/tx/detail/DelegateSet.h>
+#include <xrpld/ledger/View.h>
+
+#include <xrpl/basics/Log.h>
+#include <xrpl/protocol/Feature.h>
+#include <xrpl/protocol/Indexes.h>
+#include <xrpl/protocol/TxFlags.h>
+#include <xrpl/protocol/st.h>
+
+namespace ripple {
+
+NotTEC
+DelegateSet::preflight(PreflightContext const& ctx)
+{
+    if (!ctx.rules.enabled(featurePermissionDelegation))
+        return temDISABLED;
+
+    if (auto const ret = preflight1(ctx); !isTesSuccess(ret))
+        return ret;
+
+    auto const& permissions = ctx.tx.getFieldArray(sfPermissions);
+    if (permissions.size() > permissionMaxSize)
+        return temARRAY_TOO_LARGE;
+
+    // can not authorize self
+    if (ctx.tx[sfAccount] == ctx.tx[sfAuthorize])
+        return temMALFORMED;
+
+    std::unordered_set<std::uint32_t> permissionSet;
+
+    for (auto const& permission : permissions)
+    {
+        auto const permissionValue = permission[sfPermissionValue];
+
+        if (permissionSet.find(permissionValue) != permissionSet.end())
+            return temMALFORMED;
+
+        permissionSet.insert(permissionValue);
```

Comment:
You can just directly attempt to insert and handle the failure.. You're actually traversing the permissionSet 2 times here; once to check and again to insert.

```suggestion
        if (!permissionSet.insert(permissionValue).second)
            return temMALFORMED;
```

Reply:
Since the `permissionSet` is of type `std::unordered_set<std::uint32_t>`, it uses hash to find element instead of traversing. The time complexity will be O(1) unless the set is too large and there's hash collisions. But our permission set size can not exceed 10, so it's still constant time to do `.find()`